llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -643,8 +644,34 @@ bool RISCVVLOptimizer::isCandidate(const MachineInstr &MI) const {
 
   unsigned VLOpNum = RISCVII::getVLOpNum(Desc);
   const MachineOperand &VLOp = MI.getOperand(VLOpNum);
-  if (!VLOp.isImm() || VLOp.getImm() != RISCV::VLMaxSentinel)
+  if (((VLOp.isImm() && VLOp.getImm() != RISCV::VLMaxSentinel) ||
+       VLOp.isReg())) {
+    bool UseTAPolicy = false;
+    bool HasPassthru = RISCVII::isFirstDefTiedToFirstUse(Desc);
+    if (RISCVII::hasVecPolicyOp(Desc.TSFlags)) {
+      unsigned PolicyOpNum = RISCVII::getVecPolicyOpNum(Desc);
+      const MachineOperand &PolicyOp = MI.getOperand(PolicyOpNum);
+      uint64_t Policy = PolicyOp.getImm();
+      UseTAPolicy = (Policy & RISCVII::TAIL_AGNOSTIC) == RISCVII::TAIL_AGNOSTIC;
+      if (HasPassthru) {
+        unsigned PassthruOpIdx = MI.getNumExplicitDefs();
+        UseTAPolicy = UseTAPolicy || (MI.getOperand(PassthruOpIdx).getReg() ==
+                                      RISCV::NoRegister);
+      }
```

Comment:
I think some of the mask instruction pseudos have a passthru without a policy operand, do we need to check the passthru separately?

Reply:
Do you know what the TA policy is if we do not have a policy operand?

Reply:
Looking at RISCVInsertVSETVLI I think it's TU when there's no policy operand.

But I'm wondering now actually if it's safe to reduce the VL of a pseudo that's TA with a defined passthru, given that TA is still somewhat defined.

Would it be better to be conservative here and bail unless the passthru is undefined? I think that still covers the vast majority of cases, tail agnostic with a defined passthru isn't very common.

Reply:
addressed in the latest fixup.